File: backend/services/chat_template_service.py
import logging
from typing import List

from consts.exceptions import AgentRunException
from database import chat_template_db

logger = logging.getLogger(__name__)


async def list_templates_service(user_id: str, tenant_id: str) -> List[dict]:
    try:
        return chat_template_db.list_templates(user_id, tenant_id)
    except Exception as e:
        logger.error(f"Failed to list templates: {e}")
        raise AgentRunException(f"Failed to list templates: {e}")


async def create_template_service(data: dict, user_id: str, tenant_id: str) -> dict:
    try:
        for field in ['template_name', 'slash_command', 'prompt_template']:
            if not data.get(field):
                raise ValueError(f"Missing required field: {field}")
        result = chat_template_db.create_template(data, user_id, tenant_id)
        return {"success": True, "template_id": result["template_id"], "message": "Template created"}
    except ValueError as e:
        raise e
    except Exception as e:
        logger.error(f"Failed to create template: {e}")
        raise AgentRunException(f"Failed to create template: {e}")


async def update_template_service(template_id: int, data: dict, user_id: str, tenant_id: str) -> dict:
    try:
        success = chat_template_db.update_template(template_id, data, user_id, tenant_id)
        if not success:
            raise ValueError(f"Template {template_id} not found or not owned by user")
        return {"success": True, "message": "Template updated"}
    except ValueError as e:
        raise e
    except Exception as e:
        logger.error(f"Failed to update template: {e}")
        raise AgentRunException(f"Failed to update template: {e}")


async def delete_template_service(template_id: int, user_id: str, tenant_id: str) -> dict:
    try:
        success = chat_template_db.soft_delete_template(template_id, user_id, tenant_id)
        if not success:
            raise ValueError(f"Template {template_id} not found or not owned by user")
        return {"success": True, "message": "Template deleted"}
    except ValueError as e:
        raise e
    except Exception as e:
        logger.error(f"Failed to delete template: {e}")
        raise AgentRunException(f"Failed to delete template: {e}")

File: frontend/services/chatTemplateService.ts
import { API_ENDPOINTS } from "./api";
import { getAuthHeaders } from "@/lib/auth";
import log from "@/lib/logger";

export interface FieldDefinition {
  key: string;
  label: string;
  required: boolean;
  type: "text" | "select" | "date" | "textarea";
  options?: string[];
}

export interface ChatTemplate {
  template_id: number;
  template_name: string;
  slash_command: string;
  prompt_template: string;
  fields: FieldDefinition[];
  sort_order: number;
  user_id: string;
  tenant_id: string;
  create_time?: string;
  update_time?: string;
}

export interface CreateTemplateRequest {
  template_name: string;
  slash_command: string;
  prompt_template: string;
  fields: FieldDefinition[];
  sort_order?: number;
}

export interface UpdateTemplateRequest {
  template_name?: string;
  slash_command?: string;
  prompt_template?: string;
  fields?: FieldDefinition[];
  sort_order?: number;
}

export async function listTemplates(): Promise<ChatTemplate[]> {
  try {
    const response = await fetch(API_ENDPOINTS.chatTemplate.list, {
      method: "GET",
      headers: getAuthHeaders(),
    });
    if (!response.ok) throw new Error(`Failed to list templates: ${response.statusText}`);
    const data = await response.json();
    return data.templates;
  } catch (error) {
    log.error("Failed to list templates:", error);
    throw error;
  }
}

export async function createTemplate(
  req: CreateTemplateRequest
): Promise<{ success: boolean; template_id: number }> {
  try {
    const response = await fetch(API_ENDPOINTS.chatTemplate.create, {
      method: "POST",
      headers: { ...getAuthHeaders(), "Content-Type": "application/json" },
      body: JSON.stringify(req),
    });
    if (!response.ok) throw new Error(`Failed to create template: ${response.statusText}`);
    return await response.json();
  } catch (error) {
    log.error("Failed to create template:", error);
    throw error;
  }
}

export async function updateTemplate(
  id: number,
  req: UpdateTemplateRequest
): Promise<{ success: boolean }> {
  try {
    const response = await fetch(API_ENDPOINTS.chatTemplate.update(id), {
      method: "PUT",
      headers: { ...getAuthHeaders(), "Content-Type": "application/json" },
      body: JSON.stringify(req),
    });
    if (!response.ok) throw new Error(`Failed to update template: ${response.statusText}`);
    return await response.json();
  } catch (error) {
    log.error("Failed to update template:", error);
    throw error;
  }
}

export async function deleteTemplate(id: number): Promise<{ success: boolean }> {
  try {
    const response = await fetch(API_ENDPOINTS.chatTemplate.delete(id), {
      method: "DELETE",
      headers: getAuthHeaders(),
    });
    if (!response.ok) throw new Error(`Failed to delete template: ${response.statusText}`);
    return await response.json();
  } catch (error) {
    log.error("Failed to delete template:", error);
    throw error;
  }
}
